### PR TITLE
Ability to add and remove tasks from the event edit form

### PR DESCRIPTION
**Pull request by neon-ninja**
Solves #16

amy task edit delete

### CODE REVIEW

File: workshops/templates/base_without_navbar.html [hunk 1]
```diff
@@ -4,9 +4,12 @@
   <head>
     <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap.min.css" />
     <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.1/css/bootstrap-theme.min.css" />
+    <link rel="stylesheet" href="https://code.jquery.com/ui/1.11.4/themes/smoothness/jquery-ui.css">
```
Comment (pbanaszkiewicz) [inline workshops/templates/base_without_navbar.html]: There's one char missing at the end: `/`. The line should look like `<link rel="stylesheet" href="..." />`, otherwise it makes a lot of noise in the tests.

File: workshops/views.py [hunk 2]
```diff
@@ -441,11 +442,33 @@ class EventCreate(LoginRequiredMixin, CreateViewContext):
     model = Event
     fields = '__all__'
 
+@login_required
+def event_edit(request, event_ident):
+    event_form_factory = modelformset_factory(Event, fields='__all__', extra=0)
+    task_form_factory = inlineformset_factory(Event, Task, fields='__all__', extra=1, can_delete=False)
```
Comment (pbanaszkiewicz) [inline workshops/views.py]: I'm not certain about these two form factories. Wouldn't it be better to have ModelForms explicitely implemented in `workshops.forms`?

File: workshops/views.py [hunk 3]
```diff
@@ -474,6 +497,13 @@ def task_details(request, task_id):
     return render(request, 'workshops/task.html', context)
 
 
+@login_required
+def task_delete(request, task_id):
```
Comment (pbanaszkiewicz) [inline workshops/views.py]: Please add a docstring explaining where this view is used.

Comment (wking) [inline workshops/views.py]: On Wed, Apr 01, 2015 at 02:00:24PM -0700, Piotr Banaszkiewicz wrote:
> > +@login_required
> > +def task_delete(request, task_id):
> 
> Please add a docstring explaining where this view is used.

I'm -1 on that.  Docstring's should explain what a function (or
whatever) does, not who's using it.  It's better to pick distinct
names and use:

  $ git grep task_delete origin/pr/234

or similar to figure out who's using the function.  It's also good to
motivate odd stuff in the commit message (but this example seems
obvious enough to not need that).  And it would be good if we had some
user-facing workflow documentation, but we don't have anything along
those lines yet.

**Comment by pbanaszkiewicz** [comment]
Hey @neon-ninja,

thanks for your PR!

I found a few very minor issues in the code.

One note regarding UI/UX: when I tried this out without looking at the code, I thought it uses AJAX and would add a new task without reloading the page.

Now, after reading the code, I know that a new task is created and the event is saved, and then a redirect happens.

I think it would be better to have two submit buttons ("Submit" and "Add"). Upon clicking "Add", a new task would be created, but the event form data would be saved and restored after the redirect to the same page.
Upon clicking "Submit" only changes to the event would be saved (then redirect to `event_details` view).

What do you think about this idea, Nick?

File: workshops/views.py (hunk 3 above)
Comment (pbanaszkiewicz) [inline workshops/views.py]: A usual CRUD would redirect to the list of objects of the same kind after deleting one object from that list.

> Docstring's should explain what a function (or whatever) does, not who's using it.

In case of deleting a task, that function redirects to the task's event. The redirect is there to help with inline edits from the event's page. I think this is unusual enough to at least deserve a comment. This redirect is also a part of function's behavior, so it may just as well be put in the docstring.

> It's better to pick distinct names

Personally I'd use a different name and URL for that view - or jump to "RESTful" fashion and return a different status code instead of the redirect.

Comment (wking) [inline workshops/views.py]: On Wed, Apr 01, 2015 at 02:27:23PM -0700, Piotr Banaszkiewicz wrote:
> > Docstring's should explain what a function (or whatever) does, not
> > who's using it.
> 
> In case of deleting a task, that function redirects to the task's
> event. The redirect is there to help with inline edits from the
> event's page. I think this is unusual enough to at least deserve a
> comment. This redirect is also a part of function's behavior, so it
> may just as well be put in the docstring.

Ah, I agree that it's good to explain any surprising redirects
returned by the function it that function's docstring.  When you said
“where this view is used”, I thought you meant the URL the view was
bound to and any views linking to that URL, not where you'd go after
visiting this view.

File: workshops/views.py (hunk 2 above)
Comment (neon-ninja) [inline workshops/views.py]: Could do. The one line solution just seemed more elegant to me. Particularly inlineformset seemed to be the only way to generate a model form that uses foreign keys automatically referring to the event on the page.

**Comment by neon-ninja** [comment]
Sounds good to me :)

**Comment by pbanaszkiewicz** [comment]
Hi @neon-ninja

Thanks for your commits!

The latest flow in the event edit view is not exactly what I imagined in https://github.com/swcarpentry/amy/pull/234#issuecomment-88632603

I think there should be two buttons, for example "Save" and "Add". The former should be placed right below the event edit form, while the latter should be where it is right now - below the tasks table.

Upon clicking on "Save": Amy should attempt to save the event form, and remember the contents of the task form (without saving them to the database).

Upon clicking on "Add": Amy should attempt to add a new task, and remember the contents of the event form (without saving them to the database).

I hope my vision is now clearer :)

Cheers!

File: workshops/templates/workshops/event_form.html [hunk 4]
```diff
@@ -3,21 +3,44 @@
...
 {% block breadcrumbs %}
     {% breadcrumb_main_page %}
-    {% breadcrumb_index_all_objects model %}
-    {% if object %}
-        {% breadcrumb_object object %}
-        {% breadcrumb_edit_object object %}
-    {% else %}
-        {% breadcrumb_new_object model %}
-    {% endif %}
+    {% breadcrumb_active title %}
```
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/event_form.html]: Why did you remove the breadcrumbs? Was there an error or something?

Comment (neon-ninja) [inline workshops/templates/workshops/event_form.html]: yes, as I wasn't passing in model or object. I can, just for the sake of the breadcrumbs.

File: workshops/templates/workshops/event_form.html [hunk 5]
```diff
@@ -20,4 +20,4 @@
     <input type="submit" value="Submit" />
 </form>
 
-{% endblock %}
+{% endblock %}
```
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/event_form.html]: Please bring back the last line of this file.

**Comment by pbanaszkiewicz** [comment]
Okay, this looks nice.

Apart from [the missing last line](https://github.com/swcarpentry/amy/pull/234#discussion-diff-28274049) I have only one comment regarding redirects:

1. After clicking "Save": redirect to the event's detail view
2. After clicking "Add": leave as it is right now (ie. "refresh" the page)
3. After clicking "Delete": change to "refresh" (ie. redirect to the event edit view)

@wking do you have any comments regarding this PR? I might have missed something.

**Comment by wking** [comment]
On Mon, Apr 13, 2015 at 01:17:52PM -0700, Piotr Banaszkiewicz wrote:
> … I have only one comment regarding redirects:

Staying on the event-edit page after adding/editing/deleting tasks
makes sense to me.

> @wking do you have any comments regarding this PR?

No, it looks good to me, and your comments seem reasonable.

**Comment by neon-ninja** [comment]
@pbanaszkiewicz see above commits :)

**Comment by pbanaszkiewicz** [comment]
Hey @neon-ninja,

I took your work and made some minor changes:

1. rebase against latest master
2. change formsets to "normal" forms
3. fix errors caused by 1 & 2

I'm going to merge everything now. Thanks a lot for your contributions! :-) Very appreciated. And sorry it took me so long to review and merge this…

**Comment by pbanaszkiewicz** [comment]
Hmmm, this should've been resolved by GitHub automatically. I'm not sure why it didn't.

**Comment by narayanaditya95** [comment]
@pbanaszkiewicz I am fixing it.

**Comment by pbanaszkiewicz** [comment]
My guess is that because I rebased this PR against `master` and did not forcibly update the PR, it shows as unmerged, even though @neon-ninja's commits show in `master`.

So I'm closing this PR. Thanks again @neon-ninja for your contribution, and sorry everyone for the chaos I made due to silly rebasing.
